lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -58,13 +94,63 @@ FModuleOp LowerLayersPass::buildNewModule(OpBuilder &builder, Location location,
   return newModule;
 }
 
-/// Process a module to remove any layer blocks it has.
-void LowerLayersPass::runOnModule(FModuleOp moduleOp) {
+void LowerLayersPass::removeLayersFromValue(Value value) {
+  auto type = dyn_cast<RefType>(value.getType());
+  if (!type || !type.getLayer())
+    return;
+  value.setType(type.removeLayer());
+}
+
+void LowerLayersPass::removeLayersFromPorts(FModuleLike moduleLike) {
+  if (auto moduleOp = dyn_cast<FModuleOp>(moduleLike.getOperation())) {
+    for (auto arg : moduleOp.getBodyBlock()->getArguments())
+      removeLayersFromValue(arg);
+  }
+
+  auto oldTypeAttrs = moduleLike.getPortTypesAttr();
+  SmallVector<Attribute> newTypeAttrs;
+  newTypeAttrs.reserve(oldTypeAttrs.size());
+  for (auto typeAttr : oldTypeAttrs.getAsRange<TypeAttr>()) {
+    if (auto refType = dyn_cast<RefType>(typeAttr.getValue()))
+      if (refType.getLayer())
+        typeAttr = TypeAttr::get(refType.removeLayer());
+    newTypeAttrs.push_back(typeAttr);
+  }
+  moduleLike->setAttr(FModuleLike::getPortTypesAttrName(),
+                      ArrayAttr::get(moduleLike.getContext(), newTypeAttrs));
+}
+
+void LowerLayersPass::removeLayersFromRefCast(RefCastOp cast) {
+  auto result = cast.getResult();
+  auto oldType = result.getType();
+  if (oldType.getLayer()) {
+    auto input = cast.getInput();
+    auto srcType = input.getType();
+    auto newType = oldType.removeLayer();
+    if (newType == srcType) {
+      result.replaceAllUsesWith(input);
+      cast->erase();
+      return;
+    }
+    result.setType(newType);
+  }
+}
+
+void LowerLayersPass::runOnModule(FModuleLike moduleLike) {
   LLVM_DEBUG({
-    llvm::dbgs() << "Module: " << moduleOp.getModuleName() << "\n";
+    llvm::dbgs() << "Module: " << moduleLike.getModuleName() << "\n";
     llvm::dbgs() << "  Examining Layer Blocks:\n";
   });
 
+  // Strip away layers from the interface of the module-like op.
+  moduleLike.setLayers({});
+  removeLayersFromPorts(moduleLike);
```

Comment:
This is safe and has no effect on things like ClassOp. However, would it be better to early exit on these instead?